invenio_s3/multipart_client.py

```diff
@@ -0,0 +1,147 @@
+# To overcome this limitation, we have to use the low-level API directly separated in the
+# MultipartS3File class.
+
+
+class MultipartS3File:
+    """Low level client for S3 multipart uploads."""
+
+    def __init__(self, fs, path, upload_id=None):
+        """Initialize the low level client.
+
+        :param fs: S3FS instance
+        :param path: The path of the file (with bucket and version)
+        :param upload_id: The upload ID of the multipart upload, can be none to get a new upload.
+        """
+        self.fs = fs
+        self.path = path
+        self.bucket, self.key, self.version_id = fs.split_path(path)
+        self.s3_client = fs.s3
+        self.acl = fs.s3_additional_kwargs.get("ACL", "")
+        self.upload_id = upload_id
+
+    def create_multipart_upload(self):
+        """Create a new multipart upload.
+
+        :returns: The upload ID of the multipart upload.
+        """
+        mpu = self.s3_client.create_multipart_upload(
+            Bucket=self.bucket, Key=self.key, ACL=self.acl
+        )
+        # TODO: error handling here
+        self.upload_id = mpu["UploadId"]
+        return self.upload_id
+
+    def get_parts(self, max_parts):
+        """List the parts of the multipart upload.
+
+        :param max_parts: The maximum number of parts to list.
+        :returns: The list of parts, including checksums and etags.
+        """
+        ret = self.s3_client.list_parts(
+            Bucket=self.bucket,
+            Key=self.key,
+            UploadId=self.upload_id,
+            MaxParts=max_parts,
+            PartNumberMarker=0,
+        )
+        return ret.get("Parts", [])
+
+    def upload_part(self, part_number, data):
+        """Upload a part of the multipart upload. Will be used only in tests.
+
+        :param part_number: The part number.
+        :param data: The data to upload.
+        """
+        part = self.s3_client.upload_part(
+            Bucket=self.bucket,
+            Key=self.key,
+            UploadId=self.upload_id,
+            PartNumber=part_number,
+            Body=data,
+        )
+        return part
+
+    def _complete_operation_part_parameters(self, part):
+        """Filter parameters for the complete operation."""
+        ret = {}
+        for k in [
+            "PartNumber",
+            "ETag",
+            "ChecksumCRC32",
+            "ChecksumCRC32C",
+            "ChecksumSHA1",
+            "ChecksumSHA256",
+        ]:
+            if k in part:
+                ret[k] = part[k]
+        return ret
+
+    def get_part_links(self, max_parts, url_expiration):
+        """
+        Generate pre-signed URLs for the parts of the multipart upload.
+
+        :param max_parts: The maximum number of parts to list.
+        :param url_expiration: The expiration time of the URLs in seconds
+
+        :returns: The list of parts with pre-signed URLs and expiration times.
+        """
+        expiration = datetime.datetime.utcnow() + datetime.timedelta(
```

Comment:
Can we use `datetime.datetime.now(timezone.utc)` instead? `utcnow` has been deprecated in `3.12`. I know we are supporting since `3.9` (for now), but this change seems rather simple to make it future-proof.

Reply:
there is a problem with `datetime.datetime.now(timezone.utc)`. it is not completely equivalent to `utcnow()`. 

`datetime.now(timezone.utc).replace(tzinfo=None)` would be the equivalent.

side information: https://github.com/inveniosoftware/invenio-db/pull/178 and https://github.com/inveniosoftware/invenio-db/pull/179